PublicComponentCode/ComponentCode/src/java/main/ccsu/edu/grovepicomponents/TemperatureAndHumiditySensor.java

```diff
@@ -0,0 +1,132 @@
+package ccsu.edu.grovepicomponents;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+
+import edu.ccsu.error.IncompatibleSensorError;
+import edu.ccsu.interfaces.Iterator;
+import edu.ccsu.utility.CommonConstants;
+
+/***
+ * Temperature and humidity sensor that reads data from
+ * the GrovePi temp/humidity sensor and returns that data
+ *
+ */
+public class TemperatureAndHumiditySensor extends Sensor {
+
+	private List<TempAndHumidityData> sensorData;
```

Comment:
And now the real power of generics and visibility comes into play.  The list of sensorData (used by all concrete classes currently: TemperatureAndHumiditySensor, LightSensor) can be refactored to the parent Sensor abstract class.  You obviously still want the sensor data typed such that it is useful within the individual subclass sensor so rather than the parent class dictating what the type is specify it as a generic that the subclass will provide the type.  Note also here that your sensor data classes should be nested static classes rather than just inner classes as they don't need to be tied to the outer class they were created by.  So essentially you want this:
  >public class TemperatureAndHumiditySensor extends Sensor<TemperatureAndHumiditySensor.TempAndHumidityData>{

But wait the compiler gives me a problem if I try and give it a private class as the generic!  Turns out this is the exact reason the protected key word is a bit odd in terms of its visibility so that it doesn't just give access to subclasses, but also everything in the package.  It was created for situations like this where you want to keep the inner class protected, but allow it to be specified as a possible type of the parent class when using generics.  This is actually a rather beautiful use of all of those pattern elements :)

So to summarize, for your data classes on both classes those should be static protected classes, you should refactor the sensor data list to the Sensor class making it use Generics, for the child classes now they will extend Sensor specifying which data class they are binding to.